### DISCUSSION TITLE
7-parameter ddm in HSSM possible?

### DISCUSSION THREAD

**Discussion by Franzi2114** [lnccbrown/HSSM#255] [Q&A] [answered]
Hello together,

is it possible to estimate the full 7-parameter diffusion model in HSSM? If yes, how?

Best,
Franziska

**Comment by digicosmos86** [comment]
Hi Franziska,

It is possible through blackbox likelihoods if you are not using any hierarchical models at the moment. Happy to provide an example if that's if you want. We will be adding hierarchical model support for full_ddm models via blackbox likelihood soon.

**Reply by Franzi2114** [reply]
Hi Paul,
interesting! It would be nice to see an example how this works =D

**Reply by digicosmos86** [reply, marked as answer]
Here's [an example](https://lnccbrown.github.io/HSSM/tutorials/likelihoods/#using-blackbox-likelihoods) in our tutorial. In this example, we basically use the full HDDM likelihood in Cython as a blackbox likelihood function. You can modify this example like this to make it work for you:

```python
# You need to do pip install git+https://github.com/brown-ccv/hddm-wfpt.git first to install 
# the hddm likelihood functions.

import hddm_wfpt
import bambi as bmb

# Define a function with fun(data, *) signature
def my_blackbox_loglik(data, v, sv, a, z, sz, t, st, err=1e-8):
    data = data[:, 0] * data[:, 1]

    # Our function expects inputs as float64, but they are not guaranteed to
    # come in as such --> we type convert
    return hddm_wfpt.wfpt.pdf_array(
        np.float64(data),
        np.float64(v),
        np.float64(sv),
        np.float64(2 * a),
        np.float64(z),
        np.float64(sz),
        np.float64(t),
        np.float64(st),
        err,
        1,
    )


# Create the model with pdf_ddm_blackbox
model = hssm.HSSM(
    data=data,
    model="full_ddm",
    loglik=my_blackbox_loglik,
    loglik_kind="blackbox",
    model_config={
        "list_params": ["v", "sv", "a", "z", "sz", "t", "st"],
        "bounds": {
            "v": (-10.0, 10.0),
            "a": (0.0, 4.0),
            "z": (0.0, 1.0),
            "t": (0.0, 2.0), # Feel free to add additional bounds
        }
    },
    t=bmb.Prior("Uniform", lower=0.0, upper=2.0, initval=0.1),
)

sample = model.sample()
```

**Reply by Franzi2114** [reply]
Great, thanks! So it's also possible to set priors for all model parameters now?

**Reply by digicosmos86** [reply]
Yes, please refer to [this guide](https://lnccbrown.github.io/HSSM/getting_started/getting_started/) to specify all priors

**Comment by frankmj** [comment]
we may also add a LAN or other approximate likelihood for the full DDM
model as that is much faster than doing numerical integration (see Fengler
eLife paper) - but requires some more work to properly span the parameter
space and validate before including in HSSM.

Michael
